### ISSUE TITLE
Possibly add functions for reading/writing npy files

### ISSUE THREAD

**Issue by mars0i** [owlbarn/owl#252]
Maybe add functions to read/write npy files, or encourage use of an OCaml library for reading/writing npy files, if it can be used on Owl matrices/ndarrays without needing wrappers:
https://github.com/LaurentMazare/npy-ocaml

npy is a numpy ndarray format:
https://www.codementor.io/tirthajyotisarkar/why-you-should-start-using-npy-file-more-often-hs8dwicsq

It seems to be readable in Matlab:
https://github.com/kwikteam/npy-matlab

And in R:
http://dirk.eddelbuettel.com/code/rcpp.cnpy.html

Although there may be difficulties for reading npy files in Matlab and/or R:
https://stackoverflow.com/questions/40529100/how-can-i-open-a-3d-npy-file-in-r-or-matlab

**Comment by tachukao** [comment]
@mars0i There's an excellent package that read/write npy files to/from bigarrays https://github.com/LaurentMazare/npy-ocaml
Owl ndarrays are genarrays so it should be pretty straightforward to do the conversion

**Comment by tachukao** [comment]
@mars0i the latest PR #470  added functions to read/write npy files

**Comment by mars0i** [comment]
That's great!

Earlier there was a decision to put this sort of thing in a separate library, and I created https://github.com/owlbarn/owl_serialisation for this purpose.  That has some simple bin_prot serialization functions, but nothing else.  I have not done anything with that in a long time, and I never put it into opam.  (I never took the time to figure out how to do that.) 

Anyway, I wonder whether the npy IO should be in that library or another separate library.  I have no opinion about it.  Just raising the question.

**Comment by tachukao** [comment]
@mars0i oh cool! I wasn't aware of owl-serialisation, looks really good! It would be great to have serialisation for npy, json, and hdf5 like you said in the README.

As for whether numpy IO should be in Owl, I think `save_npy` and `load_npy` fit in quite nicely in Owl's `Ndarray` module with other saving and loading functions. But I don't think it would hurt to have similar functions in owl-serialisation with the added functionality of serialisation and deserialisation.  What do you think?

**Comment by mars0i** [comment]
@tachukao the original reason for splitting out the serialization functions was to reduce dependencies on libraries for people who didn't need those particular kinds of serialization.  I think this is up to @ryanrhymes .  He closed the issue, so maybe it make sense to leave the npy conversion functions in Owl itself.  Liang, I added my comment just in case you had forgotten about owl_serialization and would prefer that the npy functions be moved there.  I don't have feelings about it one way or another.  I can try to figure out how to submit owl_serialization to opam some time if that seems worthwhile.

**Comment by tachukao** [comment]
@mars0i Argh I see, yea that makes sense. I don't really have a strong opinion about it either. Dependency-wise, `load_npy` and `save_npy` are simple wrappers around the read/write functions in [npy-ocaml](https://github.com/LaurentMazare/npy-ocaml), which is fairly lightweight.

**Comment by mars0i** [comment]
I'll be interested to see what Liang says.  The problem with more dependencies, I feel, isn't that it makes installation bloated--maybe npy-ocaml doesn't add much size, so that might not be the issue--but just that it means that there are more things that can go wrong when installing Owl.  If someone will never need X, and they have trouble installing because of a dependency that X has, that can be frustrating.  This is all about tradeoffs, though.

**Comment by ryanrhymes** [comment]
> The problem with more dependencies, I feel, isn't that it makes installation bloated--maybe npy-ocaml doesn't add much size, so that might not be the issue--but just that it means that there are more things that can go wrong when installing Owl. 

This is true. Especially some external dependency like Plplot, they increase the risk of installation failures. Even for pure ocaml libs, sometimes opam solver can fail due to conflicting version constraints, which happened to me a couple of times when there is a version upgrade of ocaml compilers. Maybe the situation has improved a bit.

So I am very conservative in introducting exteranal dependency into owl base and owl core. For npy/npz, I have been asked by many people for very long time, so it appears to me a very needed function. @tachukao also I think it is pure ocaml implementation and has no external dependency except zip, right?

Anyway, I am open to all opinions and suggestions. If everybody find it make more sense to move it to owl-serialisation, we can certainly do that.

**Comment by tachukao** [comment]
> @tachukao also I think it is pure ocaml implementation and has no external dependency except zip, right?

I think that's right the only dependency is camlzip
https://github.com/LaurentMazare/npy-ocaml/blob/master/npy.opam

But yea I don't have a strong opinion about this. Perhaps we can wait to see if people experience installation failures and decide then? 

@ryanrhymes On a different note, I think it might be worth advertising Owl-related projects (e.g., owl-serialisation, owl-symbolic, OwlDE, etc.) in Owl's README?

**Comment by mars0i** [comment]
I certainly won't push for taking npy IO out of Owl.  I just wanted to raise the question in case it was relevant.  

The bin_prot functions that I wrote in owl-serialization are based on Jane Street libs, which is a big dependency, and I'm sure there will be more interest in npy format than in bin_prot IO.  

(Maybe with npy, there will be no need for bin_prot IO?  I wrote that  just to have some way to store data that could preserve both Owl and user-defined types, and and that would be portable across OCaml versions.  Maybe npy format can be used to address that need--I don't know enough about it.)
